File: frontend/src/api/email.ts
/**
 * /api/admin/email/* and /api/me/subscriptions — email + notifications.
 *
 * Types here mirror BE-1's pydantic shapes in
 * ``backend/backend/schemas/email.py``. The backend uses ``smtp_`` prefixed
 * field names (``smtp_host``, ``smtp_port``, …); we keep the same names
 * in the client so no renaming layer is needed.
 */

import { http } from './client';

// ---------------------------------------------------------------------------
// SMTP configuration
// ---------------------------------------------------------------------------

/**
 * Shape of ``GET /api/admin/email/smtp``.
 *
 * Per BE-1, ``smtp_password`` is NEVER returned with the real value — the
 * server substitutes the literal sentinel ``"***"`` on every read. On the
 * PUT path, echoing ``"***"`` back tells the backend to preserve whatever
 * is already stored.
 */
export interface SmtpConfig {
  enabled: boolean;
  smtp_host: string | null;
  smtp_port: number;
  smtp_username: string | null;
  smtp_password: string;
  smtp_from_address: string | null;
  smtp_from_name: string | null;
  use_tls: boolean;
  use_ssl: boolean;
  updated_at?: string | null;
  updated_by_user_id?: number | null;
}

/** Body for ``PUT /api/admin/email/smtp``. */
export type SmtpConfigIn = Omit<SmtpConfig, 'updated_at' | 'updated_by_user_id'>;

export interface SmtpTestResult {
  ok: boolean;
  error?: string | null;
}

export async function getSmtpConfig(): Promise<SmtpConfig> {
  const { data } = await http.get<SmtpConfig>('/admin/email/smtp');
  return data;
}

export async function updateSmtpConfig(payload: SmtpConfigIn): Promise<void> {
  await http.put('/admin/email/smtp', payload);
}

export async function testSmtpConfig(payload: SmtpConfigIn): Promise<SmtpTestResult> {
  const { data } = await http.post<SmtpTestResult>('/admin/email/smtp/test', payload);
  return data;
}

// ---------------------------------------------------------------------------
// Email templates
// ---------------------------------------------------------------------------

export interface EmailTemplate {
  id: number;
  event_type: string;
  locale: string;
  subject: string;
  body_html: string;
  body_text: string;
  is_system: boolean;
  updated_at?: string | null;
  updated_by_user_id?: number | null;
  /**
   * Optional per-event variable catalogue. BE-1's initial cut doesn't
   * surface this yet; the page falls back to "no variables declared"
   * when the array is absent or empty.
   */
  available_variables?: string[];
}

export interface EmailTemplatePreview {
  subject: string;
  body_html: string;
  body_text: string;
}

export interface EmailTemplateUpdateIn {
  subject: string;
  body_html: string;
  body_text: string;
}

export async function listEmailTemplates(): Promise<EmailTemplate[]> {
  const { data } = await http.get<EmailTemplate[]>('/admin/email/templates');
  return data;
}

export async function getEmailTemplate(id: number): Promise<EmailTemplate> {
  const { data } = await http.get<EmailTemplate>(`/admin/email/templates/${id}`);
  return data;
}

export async function updateEmailTemplate(
  id: number,
  payload: EmailTemplateUpdateIn,
): Promise<void> {
  await http.put(`/admin/email/templates/${id}`, payload);
}

/**
 * Render the template against a sample context. Passing `payload` lets the
 * caller preview unsaved edits without persisting them; omit to render the
 * stored row.
 */
export async function previewEmailTemplate(
  id: number,
  payload?: EmailTemplateUpdateIn,
): Promise<EmailTemplatePreview> {
  const { data } = await http.post<EmailTemplatePreview>(
    `/admin/email/templates/${id}/preview`,
    payload ?? {},
  );
  return data;
}

export async function resetEmailTemplate(id: number): Promise<void> {
  await http.post(`/admin/email/templates/${id}/reset`);
}

// ---------------------------------------------------------------------------
// Per-user subscriptions
// ---------------------------------------------------------------------------

/**
 * A single subscription row. ``project`` is null for the global default
 * (the per-event fallback applied when no project-specific row exists).
 */
export interface SubscriptionRow {
  event_type: string;
  project: string | null;
  enabled: boolean;
}

/** Body for ``PATCH /api/me/subscriptions``. */
export interface SubscriptionBulkIn {
  subscriptions: SubscriptionRow[];
}

export async function getMySubscriptions(): Promise<SubscriptionRow[]> {
  const { data } = await http.get<SubscriptionRow[]>('/me/subscriptions');
  return data;
}

export async function patchSubscriptions(
  subscriptions: SubscriptionRow[],
): Promise<void> {
  await http.patch('/me/subscriptions', { subscriptions });
}

// ---------------------------------------------------------------------------
// Anonymous unsubscribe
// ---------------------------------------------------------------------------

export interface UnsubscribeResult {
  ok: boolean;
  /**
   * Human-readable outcome. BE-1 uses this for both "unsubscribed from
   * <event>" and "token expired" messaging, so the UI surfaces it as-is.
   */
  detail?: string | null;
}

/**
 * The `?token=...` in the unsubscribe link is a signed one-shot string.
 * The endpoint itself is anonymous-accessible on the backend; we deliberately
 * call it through axios so the request picks up base URL + error handling
 * like every other page. If the browser still has a stored JWT (user happens
 * to be logged in), it is attached but ignored by the backend.
 */
export async function unsubscribeWithToken(token: string): Promise<UnsubscribeResult> {
  const { data } = await http.post<UnsubscribeResult>('/unsubscribe', null, {
    params: { token },
  });
  return data;
}

// ---------------------------------------------------------------------------
// Per-project multi-recipient list (v0.1.4)
// ---------------------------------------------------------------------------

/**
 * One :class:`ProjectNotificationRecipient` row as returned by the backend.
 *
 * The unsubscribe token is intentionally NOT exposed by the API — it lives
 * only in outgoing email footers so a one-click public link is the
 * recipient's exit path. The frontend never touches it.
 */
export interface ProjectRecipient {
  id: number;
  project: string;
  email: string;
  event_kinds: string[];
  enabled: boolean;
  added_by_user_id: number;
  created_at?: string | null;
  updated_at?: string | null;
}

/** Body for ``POST /api/projects/{project}/recipients``. */
export interface ProjectRecipientIn {
  email: string;
  event_kinds: string[];
  enabled: boolean;
}

/** Body for ``PATCH /api/projects/{project}/recipients/{id}``. Every field optional. */
export interface ProjectRecipientPatch {
  email?: string;
  event_kinds?: string[];
  enabled?: boolean;
}

export async function listProjectRecipients(
  project: string,
): Promise<ProjectRecipient[]> {
  const { data } = await http.get<ProjectRecipient[]>(
    `/projects/${encodeURIComponent(project)}/recipients`,
  );
  return data;
}

export async function addProjectRecipient(
  project: string,
  payload: ProjectRecipientIn,
): Promise<ProjectRecipient> {
  const { data } = await http.post<ProjectRecipient>(
    `/projects/${encodeURIComponent(project)}/recipients`,
    payload,
  );
  return data;
}

export async function updateProjectRecipient(
  project: string,
  id: number,
  patch: ProjectRecipientPatch,
): Promise<ProjectRecipient> {
  const { data } = await http.patch<ProjectRecipient>(
    `/projects/${encodeURIComponent(project)}/recipients/${id}`,
    patch,
  );
  return data;
}

export async function deleteProjectRecipient(
  project: string,
  id: number,
): Promise<void> {
  await http.delete(
    `/projects/${encodeURIComponent(project)}/recipients/${id}`,
  );
}

File: backend/backend/schemas/email.py
"""Pydantic DTOs for the ``/api/admin/email/*`` + ``/api/me/subscriptions``
endpoints.

The password field on :class:`SmtpConfigOut` / :class:`SmtpConfigIn` is
always masked as ``"***"`` on the way out.  On PUT, a value of ``"***"``
is treated as a sentinel meaning "keep the existing password" so the
admin UI can round-trip a read without clobbering the secret.
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# Sentinel the admin endpoint recognises as "preserve the stored password".
MASKED_PASSWORD: str = "***"


class SmtpConfigOut(BaseModel):
    """SMTP config projection returned by GET /api/admin/email/smtp."""

    model_config = ConfigDict(from_attributes=True)

    enabled: bool
    smtp_host: str | None = None
    smtp_port: int = 587
    smtp_username: str | None = None
    # Always masked; real secret never leaves the server.
    smtp_password: str = Field(default=MASKED_PASSWORD)
    smtp_from_address: str | None = None
    smtp_from_name: str | None = None
    use_tls: bool = True
    use_ssl: bool = False
    updated_at: str | None = None
    updated_by_user_id: int | None = None


class SmtpConfigIn(BaseModel):
    """Body for PUT /api/admin/email/smtp.

    A literal ``"***"`` in ``smtp_password`` means "don't overwrite the
    stored password"; any other value (including the empty string)
    replaces it.
    """

    model_config = ConfigDict(extra="forbid")

    enabled: bool = False
    smtp_host: str | None = None
    smtp_port: int = 587
    smtp_username: str | None = None
    smtp_password: str = Field(default=MASKED_PASSWORD)
    smtp_from_address: str | None = None
    smtp_from_name: str | None = None
    use_tls: bool = True
    use_ssl: bool = False


class SmtpTestResult(BaseModel):
    """Outcome of POST /api/admin/email/smtp/test."""

    ok: bool
    error: str | None = None


# ---------------------------------------------------------------------------
# Templates
# ---------------------------------------------------------------------------


class EmailTemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    event_type: str
    locale: str
    subject: str
    body_html: str
    body_text: str
    is_system: bool
    updated_at: str | None = None
    updated_by_user_id: int | None = None


class EmailTemplateUpdateIn(BaseModel):
    """Subject / body edits for an existing template.

    ``event_type`` and ``locale`` are immutable — changing them would
    let an admin accidentally overwrite a sibling row's identity.
    """

    model_config = ConfigDict(extra="forbid")

    subject: str
    body_html: str
    body_text: str


class EmailTemplatePreviewOut(BaseModel):
    subject: str
    body_html: str
    body_text: str


# ---------------------------------------------------------------------------
# Subscriptions
# ---------------------------------------------------------------------------


class SubscriptionRow(BaseModel):
    """One notification subscription row.

    ``project`` may be null to express "global default for this
    event_type".  Per-project rows shadow the global default.
    """

    model_config = ConfigDict(from_attributes=True)

    project: str | None = None
    event_type: str
    enabled: bool


class SubscriptionBulkIn(BaseModel):
    """Bulk upsert body for PATCH /api/me/subscriptions."""

    model_config = ConfigDict(extra="forbid")

    subscriptions: list[SubscriptionRow] = Field(default_factory=list)


class UnsubscribeResult(BaseModel):
    ok: bool
    detail: str | None = None


# ---------------------------------------------------------------------------
# Per-batch subscription override
# ---------------------------------------------------------------------------


class BatchEmailSubscriptionOut(BaseModel):
    """One :class:`BatchEmailSubscription` row.

    ``event_kinds`` is the parsed JSON list (already a Python ``list[str]``);
    the wire format hides the on-disk JSON encoding from the UI.
    """

    model_config = ConfigDict(from_attributes=True)

    batch_id: str
    event_kinds: list[str] = Field(default_factory=list)
    enabled: bool = True
    created_at: str | None = None
    updated_at: str | None = None


class BatchEmailSubscriptionIn(BaseModel):
    """Body for PUT /api/batches/{batch_id}/email-subscription."""

    model_config = ConfigDict(extra="forbid")

    event_kinds: list[str] = Field(default_factory=list)
    enabled: bool = True


# ---------------------------------------------------------------------------
# Per-project recipient list (multi-recipient notifications, v0.1.4)
# ---------------------------------------------------------------------------


class ProjectRecipientOut(BaseModel):
    """One :class:`ProjectNotificationRecipient` row.

    ``event_kinds`` is the parsed JSON list (already ``list[str]``);
    the wire format hides the on-disk JSON encoding from the UI.
    ``unsubscribe_token`` is intentionally NOT exposed here — only the
    backend embeds it in outgoing email footers, never returned to
    authenticated callers.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    project: str
    email: str
    event_kinds: list[str] = Field(default_factory=list)
    enabled: bool = True
    added_by_user_id: int
    created_at: str | None = None
    updated_at: str | None = None


class ProjectRecipientIn(BaseModel):
    """Body for ``POST /api/projects/{project}/recipients``.

    ``email`` is validated as RFC 5322 via Pydantic's :class:`EmailStr`
    so a UI typo can't smuggle ``"hello"`` into the recipient list.
    """

    model_config = ConfigDict(extra="forbid")

    email: EmailStr
    event_kinds: list[str] = Field(default_factory=list)
    enabled: bool = True


class ProjectRecipientPatchIn(BaseModel):
    """Body for ``PATCH /api/projects/{project}/recipients/{id}``.

    Every field is optional — callers PATCH only what they want to
    change.  ``email`` may be updated (e.g. correcting a typo) and
    re-validates as :class:`EmailStr`; the API enforces UNIQUE on
    ``(project, email)`` so a collision returns 409.
    """

    model_config = ConfigDict(extra="forbid")

    email: EmailStr | None = None
    event_kinds: list[str] | None = None
    enabled: bool | None = None
